From 249ff4bdaceae5802c7e47b3ff44c49d5ed9c90c Mon Sep 17 00:00:00 2001
From: Thomas Pedersen <thomas@cozybit.com>
Date: Thu, 14 Jun 2012 17:42:20 -0700
Subject: [PATCH] authsae: add init scripts

Now specifying "mesh_pass" in the vif config will start authsae.
---
 package/mac80211/files/lib/wifi/mac80211.sh |   71 ++++++++++++++++++++++++++-
 1 files changed, 70 insertions(+), 1 deletions(-)

diff --git a/package/mac80211/files/lib/wifi/mac80211.sh b/package/mac80211/files/lib/wifi/mac80211.sh
index 461b248..9c8a05b 100644
--- a/package/mac80211/files/lib/wifi/mac80211.sh
+++ b/package/mac80211/files/lib/wifi/mac80211.sh
@@ -1,6 +1,65 @@
 #!/bin/sh
 append DRIVERS "mac80211"
 
+mac80211_authsae_setup_base() {
+	local vif="$1"
+	local cfgfile="$2"
+
+	config_get ifname "$vif" ifname
+	config_get meshid "$vif" mesh_id
+	config_get mesh_pass "$vif" mesh_pass
+	config_get device "$vif" device
+	config_get hwmode "$device" hwmode
+	config_get channel "$device" channel
+	config_get htmode "$device" htmode
+
+	case "$htmode" in
+		HT20|HT40+|HT40-) htmode="$htmode";;
+		NOHT|none|*) htmode="none";;
+	esac
+
+	case "$hwmode" in
+		*g*) band=11g;;
+		*a*) band=11a;;
+	esac
+
+	cat >> "$cfgfile" <<EOF
+authsae:
+{
+ sae:
+  {
+    debug = 480;
+    password = "$mesh_pass";
+    group = [19, 26, 21, 25, 20];
+    blacklist = 5;
+    thresh = 5;
+    lifetime = 3600;
+  };
+ meshd:
+  {
+    meshid = "$meshid";
+    interface = "$ifname";
+    passive = 0;
+    debug = 1;
+    mediaopt = 1;
+    band = "$band";
+    channel = "$channel";
+    htmode = "$htmode";
+  };
+};
+
+EOF
+}
+
+mac80211_authsae_setup_mbss() {
+	local vif="$1"
+
+	cfgfile="/var/run/authsae-$vif.conf"
+	config_get ifname "$vif" ifname
+
+	mac80211_authsae_setup_base "$vif" "$cfgfile"
+}
+
 mac80211_hostapd_setup_base() {
 	local phy="$1"
 	local ifname="$2"
@@ -353,8 +412,12 @@ enable_mac80211() {
 				[ "$apidx" -gt 1 ] || iw phy "$phy" interface add "$ifname" type managed
 			;;
 			mesh)
+				local meshid
 				config_get mesh_id "$vif" mesh_id
-				iw phy "$phy" interface add "$ifname" type mp mesh_id "$mesh_id"
+				config_get mesh_pass "$vif" mesh_pass
+				# authsae starts the mesh otherwise
+				[ -z "$mesh_pass" ] && meshid='mesh_id="$mesh_id"'
+				iw phy "$phy" interface add "$ifname" type mp $meshid
 			;;
 			monitor)
 				iw phy "$phy" interface add "$ifname" type monitor
@@ -518,6 +581,12 @@ enable_mac80211() {
 					}
 				fi
 			;;
+			mesh)
+				config_get mesh_pass "$vif" mesh_pass
+				[ -z "$mesh_pass" ] && continue
+				mac80211_authsae_setup_mbss "$vif"
+				meshd-nl80211 -c /var/run/authsae-$vif.conf &> /var/log/authsae-$vif.log &
+			;;
 		esac
 		[ "$mode" = "ap" ] || mac80211_start_vif "$vif" "$ifname"
 	done
-- 
1.7.5.4

